keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,292 @@
+Eventually, in the future, this flag will be added to the kubeadm config API as well.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+to use the same approach also for kubeadm join; this point could be reconsidered
+in the future, by e.g. defining a method for allowing higher-level tools/users to
+define Kustomize patches using a new CRD.
+
+#### Storing and retrieving Kustomize patches for kubeadm
+
+Kustomize patches, should be preserved during the whole cluster lifecycle, mainly
+for allowing kubeadm to preserve changes during the kubeadm upgrade workflow.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+that Kustomize patches will remain stored in the `/etc/kubernetes/kustomize` or in a
+custom location specified with a CLI flag for the necessary time; this point could
+be reconsidered in the future, by e.g. defining a method for allowing higher-level
+tools/users to define Kustomize patches using a new CRD.
+
+### Risks and Mitigations
+
+_Confusion between kubeadm config API and kustomize_
+Kubeadm already offers a way to implement cluster settings, that is the kubeadm API
+and component configs. Adding a new feature for supporting “advanced configurations”
+can create confusion in the users.
+
+kubeadm maintainers should take care of making differences cristal clear in release notes
+and feature announcement:
+
+- The config API is well suited for most common use cases and it will remain the
+  recommended way to define core settings for a cluster, while “advanced configurations”
+  are designed for less common use cases.
+- The config API implicitly defines the main cluster variants the kubeadm team is
+  committed to support and monitor in the Kubernetes test grid, while instead higher-level
+  tools/user are responsible for the security, conformance, consistency, of 
+  “advanced configurations” for control-plane or kubelet settings. 
+
+_To not provide expected flexibility_
+In order to provide guarantee about kubeadm respecting “advanced configurations” during
+init, join, upgrades or single-phase execution, it is necessary to define some trade-offs
+around _what_ can be customized and _how_.
+
+Even if the proposed solution is based on the user feedback/issues, the kubeadm maintainers
+want to be sure the implemented is providing the expected level of flexibility and, in
+order to ensure that, we will wait for at least one K8s release cycle for the users to provide 
+feedback before moving forward in graduating the feature. 
+
+Similarly, the kubeadm maintainers should work with [`etcdadm`](https://github.com/kubernetes-sigs/etcdadm)
+project and [`addons`](https://github.com/kubernetes-sigs/addon-operators) project
+to ensure a consistent approach across different components.
+
+## Design Details
+
+### Test Plan
+
+Add at least a new E2E periodic test exercising  “advanced configurations” during init,
+join and upgrades.
```

Comment:
One mention, the surface we can test is really tiny as opposed to the really weird and possibly big combinations of patches that can happen in the wild... I think it's important to mention that we test that the patching is tested, not any kind of combination of patches themselves.

Reply:
done